### PR TITLE
W^X support

### PR DESCRIPTION
**Pull request by janvorli**
This change is the last part of enabling the W^X support. It adds the
actual executable allocator that handles all double mapped memory
allocations and creating the writeable mappings.

The platform specific functionality is placed in a new minipal that is
going to be a basis for future removal of Windows APIs usage from the
native runtime.

The last state of the change was tested on all the platforms that we
support using coreclr pri 1 tests with both the W^X enabled and disabled
using the COMPlus_EnableWXORX variable.

The debugger changes were tested using the managed debugger testing
suite on Windows x64, x86 and on Apple Silicon so far. Further testing
on other platforms is in progress.

There are also some fixes to the previously placed executable writer holders 
in the debugger code based on the actual managed debugger tests results.

I've left the debugger code in debugger\ee to use RWX pages for breakpoints
and singlestepping helpers.

Contributes to #50391

### CODE REVIEW

File: src/coreclr/minipal/Windows/doublemapping.cpp [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+            // Normally this will be successful
+            //
+            if (pResult != nullptr)
+            {
+                // return pResult
+                break;
+            }
+
+#ifdef _DEBUG
+            // if (ShouldInjectFaultInRange())
```
Comment (janvorli) [inline src/coreclr/minipal/Windows/doublemapping.cpp:167]: @jkotas do you think it is worth preserving this fault injection? I would need to either read the env var directly in this minipal or pass its state down from the executable allocator.

File: src/coreclr/minipal/Windows/doublemapping.cpp [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        }
+        else
+        {
+            // Try another section of memory
+            tryAddr = max(tryAddr + VIRTUAL_ALLOC_RESERVE_GRANULARITY,
+                          (BYTE*) mbInfo.BaseAddress + mbInfo.RegionSize);
+        }
+    }
+
+    // STRESS_LOG7(LF_JIT, LL_INFO100,
```
Comment (janvorli) [inline src/coreclr/minipal/Windows/doublemapping.cpp:190]: @jkotas using stresslog from here would require some clunky stuff. I don't remember ever needing this info from the stresslog in the past years. I was going to remove these, but I wanted to know your opinion first.

File: src/coreclr/inc/clrconfigvalues.h [hunk 3]
```diff
@@ -735,6 +735,10 @@ RETAIL_CONFIG_STRING_INFO(EXTERNAL_DOTNET_DiagnosticPorts, W("DiagnosticPorts"),
 RETAIL_CONFIG_STRING_INFO(INTERNAL_LTTngConfig, W("LTTngConfig"), "Configuration for LTTng.")
 RETAIL_CONFIG_DWORD_INFO(UNSUPPORTED_LTTng, W("LTTng"), 1, "If COMPlus_LTTng is set to 0, this will prevent the LTTng library from being loaded at runtime")
 
+//
+// Executable code
+//
+RETAIL_CONFIG_DWORD_INFO(EXTERNAL_EnableWXORX, W("EnableWXORX"), 1, "Enable W^X for executable memory.");
```
Comment (janvorli) [inline src/coreclr/inc/clrconfigvalues.h:741]: This will be flipped to default to 0 after the CI is green.

**Comment by janvorli** [comment]
The macOS legs are failing due to what seems to be an OOM - mmap failing to map a block of memory and the core file being about 3GB large. 
The installer legs on Windows x86 / x64 are failing in the ActivateClass test.
I am looking into those.

File: src/coreclr/utilcode/executableallocator.cpp [hunk 4]
```diff
@@ -0,0 +1,755 @@
...
+
+    CRITSEC_Holder csh(m_CriticalSection);
+
+    void* result = FindRWBlock(pRX, size);
+    if (result != NULL)
+    {
+        return result;
+    }
+
+    for (BlockRX* pBlock = m_pFirstBlockRX; pBlock != NULL; pBlock = pBlock->next)
```
Comment (janvorli) [inline src/coreclr/utilcode/executableallocator.cpp:693]: This would likely be worth optimizing using some better data structure than a linked list to better handle apps that JIT a lot of code. Testing this with aspnet plaintext benchmark, the max length of this list was about 100 and the average search steps to find the requested block was 27. The list has the most recent allocated block in the head, so for the case of allocating memory block and then filling it with code, the current way would perform the best as the block we try to map as rw would always be the first in the list. But for modifications of old allocated stuff, like patching stubs etc., we can end up doing many search steps.
However, the plan for post 6.0 is to try to eliminate as much of the patching of code as we can e.g. by splitting stubs into fixed code and writeable data, so I didn't want to optimize this prematurely.

**Comment by janvorli** [comment]
The OSX failure due to OOM is strange. On my local device, the test that fails actually allocates just about 6.5MB of executable memory and all the summed requests to map RW blocks are about 12MB. So I don't really see how the `mmap(NULL, size, PROT_NONE, MAP_JIT | MAP_ANON | MAP_PRIVATE, -1, 0);` with size 0x10000 could fail.

File: src/coreclr/inc/CrstTypes.def [hunk 5]
```diff
@@ -201,6 +201,10 @@ End
 Crst Exception
 End
 
+Crst ExecutableAllocatorLock
+    Unordered
```
Comment (jkotas) [inline src/coreclr/inc/CrstTypes.def:205]: Is it really safe to have it unordered? Unordered means prone to deadlocks.

File: src/coreclr/minipal/Windows/doublemapping.cpp (hunk 1 above)
Comment (jkotas) [reply src/coreclr/minipal/Windows/doublemapping.cpp:167]: We do not need to. The fault injection was not ever used in .NET Core as far as I know.

File: src/coreclr/minipal/Unix/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,4 @@
+add_library(coreclrminipal
+    STATIC
+    doublemapping.cpp
```
Comment (jkotas) [inline src/coreclr/minipal/Unix/CMakeLists.txt:3]: What you have here is fine, for now at least. I have been wondering whether we should build minipals like this in C to make easy to share with Mono when we need to, similar to what we are doing in libraries shims.

File: src/coreclr/vm/jitinterface.h [hunk 7]
```diff
@@ -238,15 +238,10 @@ extern "C" FCDECL2(Object*, ChkCastAny_NoCacheLookup, CORINFO_CLASS_HANDLE type,
 extern "C" FCDECL2(Object*, IsInstanceOfAny_NoCacheLookup, CORINFO_CLASS_HANDLE type, Object* obj);
 extern "C" FCDECL2(LPVOID, Unbox_Helper, CORINFO_CLASS_HANDLE type, Object* obj);
 
-#if defined(TARGET_ARM64) || defined(FEATURE_WRITEBARRIER_COPY)
```
Comment (jkotas) [inline src/coreclr/vm/jitinterface.h:241]: There is `ETW::MethodLog::StubsInitialized` method that gives names to the adhoc dynamically generated helpers in sampling traces so that they do not show up as broken stacks. We may want to start reporting the write barriers using this event.

File: src/coreclr/minipal/Windows/doublemapping.cpp (hunk 2 above)
Comment (jkotas) [reply src/coreclr/minipal/Windows/doublemapping.cpp:190]: Sounds good to me.

File: src/coreclr/inc/CrstTypes.def (hunk 5 above)
Comment (janvorli) [reply src/coreclr/inc/CrstTypes.def:205]: I haven't found a way to define the lock so that it doesn't hit the ordering asserts. If I just removed the Unordered, then I get e.g. this assert:

  Assert failure(PID 12572 [0x0000311c], Thread: 51440 [0xc8f0]): Consistency check failed: Crst Level violation: Can't take level 0 lock CrstExecutableAllocatorLock because you already holding level 0 lock CrstLoaderHeap

All I want here is a leaf lock. The methods in the allocator can be called from all kinds of places, so I don't know what locks it is going to be called under.

Comment (jkotas) [reply src/coreclr/inc/CrstTypes.def:205]: You can use `CrstLeafLock`. Or you can add `AcquiredAfter CrstLoaderHeap` for `CrstExecutableAllocatorLock`

Comment (janvorli) [reply src/coreclr/inc/CrstTypes.def:205]: LeafLock doesn't work, with it it still complains with the same assert. Fortunately, the AcquiredAfter LoaderHeap seems to work, so I'll use that one.

File: src/coreclr/minipal/Unix/CMakeLists.txt (hunk 6 above)
Comment (am11) [reply src/coreclr/minipal/Unix/CMakeLists.txt:3]: Agreed, the physical location of minipal could be `src/native/minipal` (we can also move `src/native/common/getexepath.h` there). Maybe it's easier to do it now since there is only doublemapping.cpp file and it's basically C compatible code (except for the scope resolution operator `::`). :)

**Comment by mangod9** [comment]
@tommcdon and @JulieLeeMSFT as FYI.

File: src/coreclr/inc/clrconfigvalues.h (hunk 3 above)
Comment (mangod9) [reply src/coreclr/inc/clrconfigvalues.h:741]: perhaps a better name for this? `EnableWXORX` is difficult to parse visually.. something like `EnableWriteXorEx`?

File: src/coreclr/vm/comcallablewrapper.h [hunk 8]
```diff
@@ -499,6 +499,7 @@ struct ComMethodTable
     // Accessor for the IDispatch information.
     DispatchInfo* GetDispatchInfo();
 
+#ifndef DACCESS_COMPILE
```
Comment (mangod9) [inline src/coreclr/vm/comcallablewrapper.h:502]: assume this is unrelated to w^x change?

File: src/coreclr/minipal/Unix/CMakeLists.txt (hunk 6 above)
Comment (janvorli) [reply src/coreclr/minipal/Unix/CMakeLists.txt:3]: I see the possibility of using this minipal by Mono as very remote. My vision is that it will gradually become a tiny replacement for the coreclr PAL, eventually allowing us to remove all Windowsisms from the coreclr. So I see this as a kind of inseparable part of the coreclr itself. While this double mapping stuff looks like something that Mono might possibly want, the other stuff that will go into the minipal is most likely going to be very coreclr specific. So I'd prefer keeping it in the coreclr folder until there is a reason to share it. 
As for converting it to C, we would need to prevent inlining of all the methods to be safe. Inlining C code into C++ is potentially dangerous e.g. due to c++ pointer aliasing. See https://github.com/dotnet/coreclr/pull/6801 for some details. We have converted all coreclr PAL files to c++ to avoid this issue in the past.

Comment (janvorli) [reply src/coreclr/minipal/Unix/CMakeLists.txt:3]: Actually, we could still compile the minipal as C++ when building it for coreclr even if it was written in C, to avoid any C/C++ combination issues. So please disregard my comment on the inlining.

File: src/coreclr/vm/comcallablewrapper.h (hunk 8 above)
Comment (janvorli) [reply src/coreclr/vm/comcallablewrapper.h:502]: They are related. They exclude methods that are not used by DAC and that would otherwise unnecessarily pull in the ExecutableWriterHolder into DAC.

File: src/coreclr/vm/jitinterface.h (hunk 7 above)
Comment (janvorli) [reply src/coreclr/vm/jitinterface.h:241]: That's nice, I'll add it.

File: src/coreclr/inc/clrconfigvalues.h (hunk 3 above)
Comment (janvorli) [reply src/coreclr/inc/clrconfigvalues.h:741]: Makes sense.

Comment (janvorli) [reply src/coreclr/inc/clrconfigvalues.h:741]: Renamed to `EnableWriteXorExecute`

File: src/coreclr/minipal/Unix/CMakeLists.txt (hunk 6 above)
Comment (jkotas) [reply src/coreclr/minipal/Unix/CMakeLists.txt:3]: Yes, it is fine to delay this until it is actually needed. I have mentioned this since I am regularly being asked by our leadership about what we can do share more code between CoreCLR and Mono. PAL layers are one of the easier opportunities.

File: src/coreclr/inc/CrstTypes.def [hunk 9]
```diff
@@ -201,6 +201,10 @@ End
 Crst Exception
 End
 
+Crst ExecutableAllocatorLock
+    AcquiredAfter LoaderHeap LeafLock
```
Comment (jkotas) [inline src/coreclr/inc/CrstTypes.def:205]: No lock should be nested inside LeafLock.

This means that some uses of LeafLock are not valid anymore, and should get its own level.

**Comment by alpencolt** [comment]
@janvorli does the double mapping map to the same memory and doesn't cause growth of memory consumption?
Do you have any docs or description of this feature?

**Comment by janvorli** [comment]
@alpencolt it doesn't map to the same virtual address, it uses a different one for writing and a different one for executing. However, the writeable mappings are temporary and live only for a very short period of time, so they don't result in incremental virtual memory usage growth. Their lifetime match the lifetime of the `ExecutableWriterHolder` instances that are used as stack only. 
There is no public doc on the feature yet, but you have a good point that we should have one.

**Comment by am11** [comment]
gcc leg is failing due to https://github.com/dotnet/runtime/pull/55299#discussion_r667469666.

**Comment by janvorli** [comment]
@jkotas I have addressed all the feedback (except for making the minipal C, which I'll postpone) and fixed a few issues that I've found based on the CI tests. Can you please take another look so that we can get it in for the preview 7? 
The PR is green except for the GCC leg that fails due to an unrelated old known issue in JIT code. So I am going to add one more commit that sets the W^X default to "off".

**Review by jkotas — APPROVED** [review]
